src/psyclone/psyir/transformations/scalarization_trans.py

```diff
@@ -0,0 +1,324 @@
+        :param var_accesses: The VariableAccessesInfo object containing
+                             signature.
+        :type var_accesses: :py:class:`psyclone.core.VariablesAccessInfo`
+        :returns bool: whether all the array accesses to signature use the
+                       same index, and whether the index is unmodified in
+                       the code region.
+        '''
+        array_indices = None
+        scalarizable = True
+        for access in var_accesses[signature].all_accesses:
+            if array_indices is None:
+                array_indices = access.component_indices
+            # For some reason using == on the component_lists doesn't work
+            elif array_indices[:] != access.component_indices[:]:
+                scalarizable = False
+                break
+            # For each index, we need to check they're not written to in
+            # the loop.
+            flattened_indices = list(itertools.chain.from_iterable(
+                    array_indices))
+            for index in flattened_indices:
+                # Index may not be a Reference, so we need to loop over the
+                # References
+                for ref in index.walk(Reference):
+                    sig, _ = ref.get_signature_and_indices()
+                    if var_accesses[sig].is_written():
+                        scalarizable = False
+                        break
+
+        return scalarizable
+
+    @staticmethod
+    def _check_first_access_is_write(signature, var_accesses):
+        '''
+        :param signature: The signature to check.
+        :type signature: :py:class:`psyclone.core.Signature`
+        :param var_accesses: The VariableAccessesInfo object containing
+                             signature.
+        :type var_accesses: :py:class:`psyclone.core.VariablesAccessInfo`
+        :returns bool: whether the first access to signature is a write.
+        '''
+        if var_accesses[signature].is_written_first():
+            return True
+        return False
+
+    @staticmethod
+    def _value_unused_after_loop(sig, var_accesses):
+        '''
+        :param sig: The signature to check.
+        :type sig: :py:class:`psyclone.core.Signature`
+        :param var_accesses: The VariableAccessesInfo object containing
+                             signature.
+        :type var_accesses: :py:class:`psyclone.core.VariablesAccessInfo`
+        :returns bool: whether the value computed in the loop containing
+                       sig is read from after the loop.
+        '''
+        # Find the last access of the signature
+        last_access = var_accesses[sig].all_accesses[-1].node
+        # Find the next accesses to this symbol
+        next_accesses = last_access.next_accesses()
+        for next_access in next_accesses:
+            # next_accesses looks backwards to the start of the loop,
+            # but we don't care about those accesses here.
```

Comment:
If we don't have it yet, we could benefit from a "node.is_descendant_of(node2)" that returns a bool, and I would use a while .parent recursion with an "is" comparison instead of abs_position to implement it.

Reply:
Ok - I'll add this as part of this PR if we don't have it.

Reply:
I'm going to add this, but in principal we could not - `path_from` sort of does this (you can try/except on ValueError and ignore the result).

Reply:
I see the new method, can you use it here?

Reply:
Yes - I've also added the loop as a parameter to `_value_unused_after_loop` to avoid computing it over and over.